xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,724 @@
+    // ACKed response's version_info).
+    @Nullable
+    private DiscoveryRequest pendingAckRequest;
+
+    private AdsStream(AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub) {
+      this.stub = checkNotNull(stub, "stub");
+    }
+
+    private void start() {
+      requestWriter = stub.withWaitForReady().streamAggregatedResources(this);
+    }
+
+    @Override
+    public void onNext(final DiscoveryResponse response) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          responseReceived = true;
+          String typeUrl = response.getTypeUrl();
+          if (typeUrl.equals(ADS_TYPE_URL_LDS)) {
+            handleLdsResponse(response);
+          } else if (typeUrl.equals(ADS_TYPE_URL_RDS)) {
+            handleRdsResponse(response);
+          }
+          // TODO(zdapeng): add CDS/EDS response handles.
+        }
+      });
+    }
+
+    @Override
+    public void onError(final Throwable t) {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.fromThrowable(t).augmentDescription("ADS stream [" + this + "] had an error"));
+        }
+      });
+    }
+
+    @Override
+    public void onCompleted() {
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          handleStreamClosed(
+              Status.UNAVAILABLE.withDescription("ADS stream [" + this + "] was closed"));
+        }
+      });
+    }
+
+    private void handleStreamClosed(Status error) {
+      checkArgument(!error.isOk(), "unexpected OK status");
+      if (closed) {
+        return;
+      }
+      closed = true;
+      cleanUp();
+      configWatcher.onError(error);
+      if (responseReceived || retryBackoffPolicy == null) {
+        // Reset the backoff sequence if had received a response, or backoff sequence
+        // has never been initialized.
+        retryBackoffPolicy = backoffPolicyProvider.get();
+      }
+      long delayNanos = 0;
+      if (!responseReceived) {
+        delayNanos =
+            Math.max(
+                0,
+                retryBackoffPolicy.nextBackoffNanos() - stopwatch.elapsed(TimeUnit.NANOSECONDS));
+      }
+      logger.log(Level.FINE, "{0} stream closed, retry in {1} ns", new Object[]{this, delayNanos});
+      if (delayNanos == 0) {
+        startRpcStream();
+        if (configWatcher != null) {
+          adsStream.sendLdsRequest(ldsResourceName);
+        }
+        // TODO(chengyuanzhang): send CDS/EDS requests if CDS/EDS watcher presents.
+      } else {
+        rpcRetryTimer =
+            syncContext.schedule(
+                new RpcRetryTask(), delayNanos, TimeUnit.NANOSECONDS, timeService);
+      }
+    }
+
+    private void close(Exception error) {
+      if (closed) {
+        return;
+      }
+      checkState(pendingAckRequest == null, "Severe bug: pending ACK/NACK not sent");
```

Comment:
Failing the `close` just because of a pendingAckRequest doesn't seem right. Can you wait/loop instead until the pendingAck is sent?

Reply:
These checks helps eliminate programmer's human error. We expect them to never happen. If check fails, the code has severe logic bug and we need to fix it.